src/main/java/org/gridsuite/modification/server/NetworkModificationService.java

```diff
@@ -8,75 +8,85 @@
-            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Switch " + switchId + " not found");
-        }
-
-        if (sw.isOpen() != open) {
-            sw.setOpen(open);
+    private Set<String> doModification(Network network, Runnable modification, NetworkModificationException.Type typeIfError) {
+        try {
+            DefaultNetworkStoreListener listener = new DefaultNetworkStoreListener();
+            network.addListener(listener);
+            modification.run();
 
             networkStoreService.flush(network);
 
-            return Set.of(sw.getVoltageLevel().getSubstation().getId());
-        } else {
-            return Set.of();
+            return listener.getModifications();
+        } catch (Exception e) {
+            NetworkModificationException exc = new NetworkModificationException(typeIfError, e.getMessage());
+            LOGGER.error(exc.getMessage());
+            throw exc;
         }
     }
 
-    public GroovyScriptResult applyGroovyScript(UUID networkUuid, String groovyScript) {
-        CompilerConfiguration conf = new CompilerConfiguration();
-        Network network = getNetwork(networkUuid);
-        DefaultNetworkStoreListener listener = new DefaultNetworkStoreListener();
-        network.addListener(listener);
-
-        Binding binding = new Binding();
-        binding.setProperty("network", network);
-        GroovyShell shell = new GroovyShell(binding, conf);
+    private Mono<Network> getNetwork(UUID networkUuid) {
         try {
-            shell.evaluate(groovyScript);
-            networkStoreService.flush(network);
-
-            return new GroovyScriptResult(true, listener.getModifications());
-        } catch (Exception ignored) {
-            return new GroovyScriptResult(false, listener.getModifications());
+            return Mono.just(networkStoreService.getNetwork(networkUuid));
+        } catch (PowsyblException e) {
+            return Mono.error(new NetworkModificationException(NETWORK_NOT_FOUND, networkUuid.toString()));
         }
     }
 
+    Mono<Set<String>> changeSwitchState(UUID networkUuid, String switchId, boolean open) {
+        return getNetwork(networkUuid)
+                .filter(network -> network.getSwitch(switchId) != null)
+                .switchIfEmpty(Mono.error(new NetworkModificationException(SWITCH_NOT_FOUND, switchId)))
+                .filter(network -> network.getSwitch(switchId).isOpen() != open)
+                .map(network -> doModification(network, () -> network.getSwitch(switchId).setOpen(open)))
+                .switchIfEmpty(Mono.just(Set.of()));
+    }
+
+    private Mono<Void> assertGroovyScriptNotEmpty(String groovyScript) {
+        return StringUtils.isBlank(groovyScript) ? Mono.error(new NetworkModificationException(GROOVY_SCRIPT_EMPTY)) : Mono.empty();
+    }
+
+    Mono<Set<String>> applyGroovyScript(UUID networkUuid, String groovyScript) {
+        return Mono.when(assertGroovyScriptNotEmpty(groovyScript)).then(
```

Comment:
why Mono.when ?

Reply:
Removed
At the beginning there were several assertions